Pytmc fails on arrays of pointers to motors

Desired behavior is to skip these kinds of objects. This one specifically is `ARRAY [0..18] OF POINTER TO DUT_MotionStage`. I'm happy to work on this tomorrow, just documenting for now.

From the bleeding edge LFE motion PLC:
```
$ make
-----------------------------------------------------------
Build path: /reg/neh/home4/zlentz/github/lcls-plc-lfe-motion/iocBoot/ioc-lfe-motion/.pytmc_build
PLC: lfe_motion
IOC name: ioc-lfe-motion
PYTMC_OPTS: 
PROJECT_PATH: ../../plc-lfe-motion/plc-lfe-motion.tsproj
Absolute project path:  /reg/neh/home4/zlentz/github/lcls-plc-lfe-motion/plc-lfe-motion/plc-lfe-motion.tsproj
-----------------------------------------------------------


mkdir -p "/reg/neh/home4/zlentz/github/lcls-plc-lfe-motion/iocBoot/ioc-lfe-motion/.pytmc_build"
Executing pytmc...
cd "/reg/neh/home4/zlentz/github/lcls-plc-lfe-motion/iocBoot/ioc-lfe-motion/.pytmc_build" && \
	pytmc stcmd \
	--name "ioc-lfe-motion" \
	--plc "lfe_motion" \
	--template-path "/reg/g/pcds/epics/ioc/common/ads-ioc/R0.1.6/iocBoot/templates" \
	--template "st.cmd.template" \
	--hashbang "/reg/g/pcds/epics/ioc/common/ads-ioc/R0.1.6/bin/rhel7-x86_64/adsIoc" \
	-p "PLC:LFE:MOTION" \
	 \
	"/reg/neh/home4/zlentz/github/lcls-plc-lfe-motion/plc-lfe-motion/plc-lfe-motion.tsproj" > st.cmd
INFO:pytmc.bin.stcmd:Found 1705 additional records; writing to lfe_motion.db
Traceback (most recent call last):
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/environment.py", line 411, in getitem
    return obj[argument]
TypeError: 'Symbol_DUT_MotionStage' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/bin/pytmc", line 11, in <module>
    load_entry_point('pytmc==2.5.0', 'console_scripts', 'pytmc')()
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/pytmc/bin/pytmc.py", line 92, in main
    func(**kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/pytmc/bin/stcmd.py", line 293, in main
    raise stashed_exception
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/pytmc/bin/stcmd.py", line 286, in main
    rendered = template.render(**template_args)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/asyncsupport.py", line 76, in render
    return original_render(self, *args, **kwargs)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/_compat.py", line 37, in reraise
    raise value.with_traceback(tb)
  File "/reg/g/pcds/epics/ioc/common/ads-ioc/R0.1.6/iocBoot/templates/st.cmd.template", line 90, in top-level template code
    {% for motor in symbols.Symbol_DUT_MotionStage | sort(attribute='nc_axis.axis_number') %}
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/filters.py", line 278, in do_sort
    return sorted(value, key=key_func, reverse=reverse)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/filters.py", line 77, in attrgetter
    item = environment.getitem(item, part)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/jinja2/environment.py", line 420, in getitem
    return getattr(obj, attr)
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/pytmc/parser.py", line 946, in nc_axis
    link = self.nc_to_plc_link
  File "/reg/g/pcds/pyps/conda/py36/envs/pcds-3.0.0/lib/python3.6/site-packages/pytmc/parser.py", line 938, in nc_to_plc_link
    raise RuntimeError(f'No NC link to DUT_MotionStage found for '
RuntimeError: No NC link to DUT_MotionStage found for 'PRG_AT2L0_SOLID.fbAT2L0_axes'
make: *** [build] Error 1
```

--- a/pytmc/bin/stcmd.py
+++ b/pytmc/bin/stcmd.py
@@ -213,7 +213,6 @@ def main(tsproj_project, *, name=None, prefix=None,
     template = jinja_env.get_template(template_filename)
 
     project = parse(tsproj_project)
-    symbols = separate_by_classname(project.find(Symbol))
 
     additional_db_files = []
     try:

--- a/pytmc/bin/stcmd.py
+++ b/pytmc/bin/stcmd.py
@@ -230,6 +230,8 @@ def main(tsproj_project, *, name=None, prefix=None,
                                f'Projects: {list(by_name)}')
 
     symbols = separate_by_classname(plc.find(Symbol))
+    motors = [mot for mot in symbols['Symbol_DUT_MotionStage']
+              if not mot.is_pointer]
 
     if not only_motor:
         other_records, _ = db.process(plc.tmc, dbd_file=dbd)
@@ -277,6 +279,7 @@ def main(tsproj_project, *, name=None, prefix=None,
 
         additional_db_files=additional_db_files,
         symbols=symbols,
+        motors=motors,
         nc=nc,
     )
 

--- a/pytmc/bin/summary.py
+++ b/pytmc/bin/summary.py
@@ -172,7 +172,7 @@ def summary(tsproj_project, use_markdown=False, show_all=False,
                 symbols = list(plc.find(parser.Symbol))
                 for symbol in sorted(symbols, key=lambda symbol: symbol.name):
                     info = symbol.info
-                    print('    {name} : {qualified_type_name} ({bit_offs} '
+                    print('    {name} : {summary_type_name} ({bit_offs} '
                           '{bit_size})'.format(**info))
                 print()
 

--- a/pytmc/parser.py
+++ b/pytmc/parser.py
@@ -623,7 +623,7 @@ class ArrayInfo(_TmcItem):
 
         ubound = (int(self.UBound[0].text)
                   if hasattr(self, 'UBound')
-                  else lbound + elements)
+                  else lbound + elements - 1)
 
         self.bounds = (lbound, ubound)
         self.elements = elements

--- a/pytmc/parser.py
+++ b/pytmc/parser.py
@@ -897,6 +897,12 @@ class Symbol(_TmcItem):
                    for owner, var in link.link):
                 yield link
 
+    @property
+    def is_pointer(self):
+        type_ = self.BaseType[0]
+        pointer_info = type_.attributes.get("PointerTo", None)
+        return bool(pointer_info)
+
 
 class Symbol_DUT_MotionStage(Symbol):
     '[TMC] A customized Symbol, representing only DUT_MotionStage'

--- a/pytmc/parser.py
+++ b/pytmc/parser.py
@@ -871,6 +871,9 @@ class Symbol(_TmcItem):
                     qualified_type_name=self.qualified_type_name,
                     bit_offs=self.BitOffs[0].text,
                     module=self.module.name,
+                    is_pointer=self.is_pointer,
+                    array_bounds = self.array_bounds,
+                    summary_type_name=self.summary_type_name,
                     )
 
     def walk(self, condition=None):
@@ -884,6 +887,18 @@ class Symbol(_TmcItem):
             return self.ArrayInfo[0]
         except (AttributeError, IndexError):
             return None
+
+    @property
+    def array_bounds(self):
+        array_info = self.array_info
+        if array_info:
+            lower_bound = int(array_info.children.LBound[0].text)
+            n_elems = int(array_info.children.Elements[0].text)
+            upper_bound = lower_bound + n_elems - 1
+            return (lower_bound, upper_bound)
+        else:
+            return None
+
 
     def get_links(self, *, strict=False):
         sym_name = '^' + self.name.lower()
@@ -902,6 +917,16 @@ class Symbol(_TmcItem):
         type_ = self.BaseType[0]
         pointer_info = type_.attributes.get("PointerTo", None)
         return bool(pointer_info)
+
+    @property
+    def summary_type_name(self):
+        summary = self.qualified_type_name
+        if self.is_pointer:
+            summary = 'POINTER TO ' + summary
+        array_bounds = self.array_bounds
+        if array_bounds:
+            summary = 'ARRAY[{}..{}] OF '.format(*array_bounds) + summary
+        return summary
 
 
 class Symbol_DUT_MotionStage(Symbol):

--- a/pytmc/parser.py
+++ b/pytmc/parser.py
@@ -872,7 +872,7 @@ class Symbol(_TmcItem):
                     bit_offs=self.BitOffs[0].text,
                     module=self.module.name,
                     is_pointer=self.is_pointer,
-                    array_bounds = self.array_bounds,
+                    array_bounds=self.array_bounds,
                     summary_type_name=self.summary_type_name,
                     )
 
@@ -898,7 +898,6 @@ class Symbol(_TmcItem):
             return (lower_bound, upper_bound)
         else:
             return None
-
 
     def get_links(self, *, strict=False):
         sym_name = '^' + self.name.lower()

--- a/pytmc/parser.py
+++ b/pytmc/parser.py
@@ -890,13 +890,9 @@ class Symbol(_TmcItem):
 
     @property
     def array_bounds(self):
-        array_info = self.array_info
-        if array_info:
-            lower_bound = int(array_info.children.LBound[0].text)
-            n_elems = int(array_info.children.Elements[0].text)
-            upper_bound = lower_bound + n_elems - 1
-            return (lower_bound, upper_bound)
-        else:
+        try:
+            return self.array_info.bounds
+        except (AttributeError, IndexError):
             return None
 
     def get_links(self, *, strict=False):

--- a/pytmc/templates/stcmd_default.cmd
+++ b/pytmc/templates/stcmd_default.cmd
@@ -24,11 +24,11 @@ epicsEnvSet("IPPORT",        "{{plc_ads_port}}")
 
 adsAsynPortDriverConfigure("$(ASYN_PORT)","$(IPADDR)","$(AMSID)","$(IPPORT)", 1000, 0, 0, 50, 100, 1000, 0)
 
-{% if symbols.Symbol_DUT_MotionStage %}
+{% if motors %}
 epicsEnvSet("MOTOR_PORT",    "{{motor_port}}")
 epicsEnvSet("PREFIX",        "{{prefix}}{{delim}}")
-epicsEnvSet("ECM_NUMAXES",   "{{symbols.Symbol_DUT_MotionStage|length}}")
-epicsEnvSet("NUMAXES",       "{{symbols.Symbol_DUT_MotionStage|length}}")
+epicsEnvSet("ECM_NUMAXES",   "{{motors|length}}")
+epicsEnvSet("NUMAXES",       "{{motors|length}}")
 
 EthercatMCCreateController("$(MOTOR_PORT)", "$(ASYN_PORT)", "$(NUMAXES)", "200", "1000")
 
@@ -57,7 +57,7 @@ asynSetTraceInfoMask("$(ASYN_PORT)", -1, 5)
 #define AMPLIFIER_ON_FLAG_WHEN_HOMING  2
 #define AMPLIFIER_ON_FLAG_USING_CNEN   4
 
-{% for motor in symbols.Symbol_DUT_MotionStage | sort(attribute='nc_axis.axis_number') %}
+{% for motor in motors | sort(attribute='nc_axis.axis_number') %}
 epicsEnvSet("AXIS_NO",         "{{motor.nc_axis.axis_number}}")
 epicsEnvSet("MOTOR_PREFIX",    "{{motor|epics_prefix}}")
 epicsEnvSet("MOTOR_NAME",      "{{motor|epics_suffix}}")

